Step1 - 엔티티 매핑

mins99 opened the pull request:
안녕하세요 리뷰어님! 처음 뵙겠습니다. 김민성입니다 : )
step1 엔티티 매핑 구현 완료하여 PR 요청드립니다.
JPA에 대해 공부한적은 있지만 실제로 현업에서 사용해 본적이 없어서
현업에서 사용하는 방식을 중점적으로 배워보려고 합니다 🧐

구현하면서 몇가지 궁금한 점이 있었습니다!

1. `findById` 의 결과를 `Optional<Answer>` 로 받을 수도 있고 `orElseThrow`나 `ifPresent`로 처리할 수도 있는데, 어떤 경우에 각각의 케이스를 써야하는지 궁금합니다! 
2. `QuestionRepository`의 `findByDeletedFalse` 테스트와 관련된 이슈가 하나 있었는데요, 전체 테스트에서 `question.setDeleted(true);` 값을 설정하게 되면 다음 순서의 테스트는 question 객체에 대한 insert 쿼리가 `deleted = true`로 들어가는 이슈가 있었습니다. 해결을 못해서 우선 해당 로직을 사용하는 테스트는 삭제하고 구현했는데, 혹시 관련해서 제가 빼먹은 설정이 있을까요?

질문이 이해가 안가시면 DM으로 편하게 물어봐주세요~!
추가 코멘트로 달려고 했었는데 해당 코멘트로 충분히 작성이 될 것 같아 따로 추가하지 않았습니다.. 
코멘트가 길어졌네요.. 무튼 JPA 미션 잘 부탁드립니다!

mins99 committed:
test(Station, Line): Spring Data JPA 실습

mins99 committed:
test(Station, Line): 다대일 단방향 연관 관계 실습

mins99 committed:
test(Station, Line): 양방향 연관 관계 실습

mins99 committed:
test(Favorite, Member): 일대다 단방향 연관 관계 실습

mins99 committed:
test(Station, LineStation): 일대일 연관 관계 실습

- 주 테이블에 외래 키(단방향, 양방향)
- 대상 테이블에 외래 키(단방향은 불가능, 양방향)

mins99 committed:
test(All): JPA hands-on 강의 내용 정리

- 파라미터 조회 속성 추가, 색상 옵션 추가
- 패키지 study -> subway 변경
- 불변 추가

mins99 committed:
docs(README): 1단계 - 엔티티 매핑 요구 사항 정리

mins99 committed:
refactor(repository): 도메인과 리포지토리 분리

mins99 committed:
refactor(application.properties): h2 DB -> MySQL Dialect 사용하도록 변경

mins99 committed:
feat(BaseTimeEntity): 공통 날짜 정보를 가지는 @MappedSuperclass 추가

- 등록일자, 수정일자

mins99 committed:
feat(All): 엔티티 클래스 작성

- DDL 요구사항에 맞춘 컬럼 지정
- 매개변수가 없는 생성자 추가

mins99 committed:
refactor(All): 설정 관련 이슈 해결

- Mysql Dialect 설정 삭제
- 지하철 실습 내용과 qna 충돌로 data.sql 삭제 및 beforeeach 추가

mins99 committed:
feat(Application): Auditing 활성화

mins99 committed:
test(Answer): domain & repository 테스트 추가

mins99 committed:
test(DeleteHistory): domain & repository 테스트 추가

mins99 committed:
test(Question): domain & repository 테스트 추가

mins99 committed:
test(User): domain & repository 테스트 추가

mins99 committed:
style(All): 코드 컨벤션 체크

mins99 committed:
docs(README): 1단계 - 엔티티 매핑 구현 사항 체크

seokhongkim commented on src/main/java/qna/Application.java:
```diff
@@ -2,7 +2,9 @@
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.data.jpa.repository.config.EnableJpaAuditing;
 
+@EnableJpaAuditing
```
JPA Audit 기능을 학습하셨군요! 👍

seokhongkim commented on src/main/java/qna/domain/BaseTimeEntity.java:
```diff
@@ -0,0 +1,22 @@
+import org.springframework.data.annotation.LastModifiedDate;
+import org.springframework.data.jpa.domain.support.AuditingEntityListener;
+
+@MappedSuperclass
```
`MappedSuperclass` 를 사용하여 공통 속성을 분리하신 것 좋습니다. 👍

seokhongkim commented on src/main/java/qna/domain/BaseTimeEntity.java:
```diff
@@ -0,0 +1,22 @@
+@EntityListeners(AuditingEntityListener.class)
+public abstract class BaseTimeEntity {
+
+    @CreatedDate
```
Audit 기능을 이용하는 `@CreatedDate`, `@LastModifiedDate` 의 경우 시간 값이 어떤 값으로 저장될까요?
한 트랜잭션 내부에서 여러개의 엔티티를 저장/수정하는 경우 각 엔티티의 생성/수정 시각은 동일한 값일까요? 다른 값일까요?
어떤 결과를 기대하고 사용하시는지 궁금합니다!
이를 기준으로 어떤 환경에서 `@CreatedDate`, `@LastModifiedDate` 를 사용하는 게 좋고, 안 좋을지 결정하실 수 있을 것 같아요!

mins99 replied:
한 트랜잭션 내부에서 여러개의 엔티티를 저장/수정하는 경우 각 엔티티별로 다른 시각이 저장됩니다!
계속 고민해봤는데 어떤 경우에 사용하는게 안좋은지는 잘 모르겠네요 ㅠㅠ 😭

seokhongkim replied:
임의의 쇼핑몰 시스템을 상상하고, 아래와 같은 경우를 고민해보시면 좋을 것 같아요! 😀 

주문이 발생하면 주문, 주문한 상품, 결제 정보 등이 저장될 수 있겠죠.
이때 주문, 주문한 상품, 결제 정보 데이터가 생성된다면 생성 일시는 동일해야 할까요? 다른 값이어도 될까요?
시스템 규모가 커지며 샤딩 또는 파티셔닝이 된다면, 데이터의 생성/수정 시각은 어떻게 맞춰줘야 할까요? 데이터 분석을 위해 주기적으로 별도의 시스템으로 데이터를 덤프해가는 시스템이 있고, 여기에서 생성/수정 시각을 기준으로 데이터를 옮긴다면 시각이 다를 때 무슨 일이 생길 수 있을까요?

seokhongkim commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -2,14 +2,39 @@
     private Long deletedById;
+
+    @Column(name = "create_date")
     private LocalDateTime createDate = LocalDateTime.now();
```
여기서는 `@CreatedDate` 를 사용하시지 않고, 값을 설정해주신 이유가 있을까요?

mins99 replied:
`@CreatedDate` 를 BaseTimeEntity 에서만 사용하는줄 알았네요 ㅎㅎ; 수정했습니다!

seokhongkim replied:
제 생각에는 `CreatedDate` 를 사용하신다면 `LocalDateTime.now()` 는 불필요할 것 같은데, 민성님은 어떻게 생각하시나요?

mins99 replied:
의도에 따라 필요할 수도 있을거는 같은데 생성자에서 createDate를 할당하도록 되어 있으니 삭제하는게 맞는것 같아요! 삭제했습니다

seokhongkim commented on src/main/java/qna/domain/User.java:
```diff
@@ -1,19 +1,32 @@
     private String email;
 
-    private User() {
+    protected User() {
```
생성자의 접근 범위를 제한하신 것 좋습니다. 👍

seokhongkim commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -19,8 +44,12 @@ public DeleteHistory(ContentType contentType, Long contentId, Long deletedById,
 
     @Override
     public boolean equals(Object o) {
```
`equals` 가 `DeleteHistory` 에 대해서만 오버라이드 되어있네요.

1. 엔티티의 `equals` 는 어떤 방법으로 구현하면 좋을까요?
2. 다른 엔티티는 `equals` 를 오버라이드 할 필요가 없을까요?

hint: `jpa entity equals` 키워드로 학습해보시면 도움이 될 것 같습니다!

mins99 replied:
JPA entity equals에 대해 찾아보니
1. 재정의 하지 않고 사용하거나
2. 데이터베이스 키를 사용하는 equals를 추가하거나
3. 비즈니스 키를 사용하는 equals를 추가하는 방법으로 사용하고 있습니다

저는 본의 아니게 1번을 실천했는데 3번이 좋을것 같기도 합니다!
[참고](https://www.baeldung.com/jpa-entity-equality)

seokhongkim replied:
학습하신 것처럼 여러 방법이 있습니다! 각각의 장단점이 있을 것이므로 팀에서, 혹은 개인적인 기준을 정해보시면 도움이 될 것 같아요 😀

seokhongkim reviewed (changes requested):
민성님 안녕하세요!
JPA 미션을 함께 할 김석홍입니다. 잘 부탁드립니다. 🙏 

먼저 질문 주신 부분에 대해 답변 드립니다.
> findById 의 결과를 Optional<Answer> 로 받을 수도 있고 orElseThrow나 ifPresent로 처리할 수도 있는데, 어떤 경우에 각각의 케이스를 써야하는지 궁금합니다!

`orElseThrow`, `ifPresent` 모두 `Optional` 로 받은 후 메소드 체이닝으로 발생하는 것이라, `Optional` 로 받는다는 것은 동일한 것으로 보입니다.
`Optional` 로 가져온 이후의 로직은 비즈니스 요구사항에 따라 달라지는 부분이므로, `orElseThrow`, `ifPresent` 는 필요에 따라 구분해서 사용하시면 될 것 같습니다.

> QuestionRepository의 findByDeletedFalse 테스트와 관련된 이슈가 하나 있었는데요, 전체 테스트에서 question.setDeleted(true); 값을 설정하게 되면 다음 순서의 테스트는 question 객체에 대한 insert 쿼리가 deleted = true로 들어가는 이슈가 있었습니다. 해결을 못해서 우선 해당 로직을 사용하는 테스트는 삭제하고 구현했는데, 혹시 관련해서 제가 빼먹은 설정이 있을까요?

`QuestionTest` 에 존재하는 `Q1`, `Q2` 의 `deleted` 값을 변경하는 테스트가 있었던 걸까요?
위 이슈라면, 각 테스트가 실행될 때 `Q1` 과 `Q2` 가 매번 초기화되는지, 이전 테스트에서 변경한 값이 그대로 유지되는지 확인해보시면 답을 찾으실 수 있을 것 같습니다!
매번 초기화되지 않는다면, 각 테스트마다 어떻게 값을 초기화해줄 수 있을지도 찾아보시면 도움이 될 것 같네요!

hint: `Q1` 과 `Q2` 는 정적 변수로 보이네요. Junit5의 `@BeforeEach` 는 어떤 용도로 쓸 수 있을까요?

이외에 코드에 피드백을 조금 남겼습니다.
지하철, 연관관계는 이번 단계의 범위가 아니므로 별도의 피드백을 남기지 않았습니다. 참고 부탁드려요!

mins99 committed:
test(All): static 변수를 참조하는 부분에 대해 로직 수정

mins99 committed:
fix(DeleteHistory): 생성일자 컬럼 수정

mins99 committed:
feat(All): equals, hashCode 추가

- JPA entity equals는 다음과 같은 방법을 사용한다
1) 재정의 하지 않거나
2) 기본키로 비교하거나
3) 비즈니스키로 비교하거나

mins99 commented:
안녕하세요 리뷰어님~! 피드백 반영하여 PR 요청 드립니다 😀
말씀주신 힌트를 통해 `Answer answer = answerRepository.save(AnswerTest.A1);` 과 같은 로직에서 
answer와 AnswerTest.A1가 동일한 객체인 것을 확인할 수 있었습니다! 감사합니다ㅠㅠ

추가로 `@BeforeEach`에서 단순히 객체만 선언을 할까, repository.save 까지 추가할까 고민이 되었는데 우선은 추후에 Mockito를 적용하는 듯 해서 객체 생성만 진행했습니다

JPA도 어렵고 TDD까지 적용하려니 고민 되는 부분이 많네요 ㅠㅠ 계속 헤매는 느낌입니다 😵‍💫
피드백 기다리면서 열심히 공부하고 있겠습니다 T_T

mins99 commented on src/test/java/qna/repository/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    @Test
+    void update() {
+        // given
+        final Answer expected = answerRepository.save(answer);
+        final String contents = expected.getContents();
+
+        // when
+        expected.setContents(contents + " new");
+        final Answer actual = answerRepository.findById(expected.getId()).orElseThrow(IllegalArgumentException::new);
+
+        // then
+        assertThat(actual.getContents()).isNotEqualTo(contents);
+    }
```
repository 각각에 대해 CRUD 테스트를 진행하는 것이 좋은가요?
좋다면 update의 경우 값이 실제로 변경된 것을 테스트 하는 것이 좋은지, 변경이 감지되었는지를 테스트하는 것이 좋은지 궁금합니다

seokhongkim replied:
민성님은 레포지토리, 그리고 CRUD 테스트를 어떤 목적으로 작성하시나요?
테스트를 작성하는 것 자체는 목적이 아니므로 민성님의 테스트 작성 의도를 생각해보시면 도움이 될 것 같습니다!

seokhongkim reviewed (approved):
민성님, JPA 1단계 미션 잘 진행해주셨습니다. 👍 

피드백 드린 부분에 대한 반영, 학습도 충분히 하신 것 같아요!
1단계 PR은 머지하도록 하겠습니다.

질문주신 부분에 대해서 코멘트를 남겼는데, 추가로 이야기 나누고 싶으신 부분이 있다면 깃헙 코멘트나 슬랙 DM으로 말씀해주세요!
2단계 미션에서 뵙겠습니다. 🙏
